Combine Base64DecodeOptions and Base64DecodeMode

annevk opened the pull request:
#### 252a605f349215717db4711cc3bfb7b7ea3bf446
<pre>
Combine Base64DecodeOptions and Base64DecodeMode
<a href="https://bugs.webkit.org/show_bug.cgi?id=255479">https://bugs.webkit.org/show_bug.cgi?id=255479</a>
rdar://108082420

Reviewed by Darin Adler.

<a href="https://bugs.webkit.org/show_bug.cgi?id=184529#c11">https://bugs.webkit.org/show_bug.cgi?id=184529#c11</a> gave a bunch of feedback that was unfortunately
overlooked. This attempts to address it. In particular:

* Using WebKit&apos;s isUnicodeCompatibleASCIIWhitespace was an accident and not intentional. All callers
  want isASCIIWhitespace (which excludes U+000B (\v)).
* The isLatin1() check was thought to be needed to address a test, but that wasn&apos;t actually the
  case.

Additionally this folds Base64DecodeOptions into Base64DecodeMode reducing the number of decoding
variants we offer to callers and making them more explicit.

We might be able to shrink that even more with some additional investigation, in particular as part
of properly standardizing CSP and SRI base64/base64url requirements.

Base64EncodeMap is renamed to Base64EncodeMode for parity.

* Source/JavaScriptCore/jsc.cpp:
(JSC_DEFINE_HOST_FUNCTION):
* Source/WTF/wtf/text/Base64.cpp:
(WTF::base64EncodeInternal):
(WTF::base64Encode):
(WTF::base64EncodeToVector):
(WTF::base64EncodeToString):
(WTF::base64DecodeInternal):
(WTF::base64Decode):
* Source/WTF/wtf/text/Base64.h:
(WTF::base64EncodeToVector):
(WTF::base64EncodeToString):
(WTF::base64Decode):
(WTF::base64URLEncodeToVector):
(WTF::base64URLEncodeToString):
(WTF::base64URLDecode):
(WTF::calculateBase64EncodedSize):
(WTF::base64Encoded):
(WTF::base64URLEncoded):
(WTF::StringTypeAdapter&lt;Base64Specification&gt;::writeTo const):
* Source/WebCore/page/Base64Utilities.cpp:
(WebCore::Base64Utilities::atob):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::userStyleSheetLocationChanged):

<a href="https://bugs.webkit.org/show_bug.cgi?id=255442">https://bugs.webkit.org/show_bug.cgi?id=255442</a> covers cleaning this up further, but this at least
aligns it with data: URL requirements.

* Source/WebCore/platform/graphics/avfoundation/CDMFairPlayStreaming.cpp:
(WebCore::extractSinfData):
* Source/WebCore/platform/network/DataURLDecoder.cpp:
(WebCore::DataURLDecoder::decodeSynchronously):
* Source/WebKit/UIProcess/Inspector/gtk/RemoteWebInspectorUIProxyGtk.cpp:
(WebKit::RemoteWebInspectorUIProxy::platformSave):
* Source/WebKit/UIProcess/Inspector/gtk/WebInspectorUIProxyGtk.cpp:
(WebKit::WebInspectorUIProxy::platformSave):
* Source/WebKit/UIProcess/Inspector/mac/WebInspectorUIProxyMac.mm:
(WebKit::WebInspectorUIProxy::showSavePanel):
* Source/WebKitLegacy/mac/WebCoreSupport/WebInspectorClient.mm:
(WebInspectorFrontendClient::save):

Canonical link: <a href="https://commits.webkit.org/263298@main">https://commits.webkit.org/263298@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c35c087de3e7147a9f7125e4b52844d385b63903

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/4161 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/4277 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/4395 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/5627 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/4408 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/4152 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/4380 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/4242 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/4631 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/4222 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/4398 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3753 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/5620 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1894 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3730 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/5935 "1 flakes 141 failures") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/3457 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3721 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/3820 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/5311 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/3952 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/4198 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/3402 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/4256 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3711 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3728 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/1043 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1024 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/7830 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/4352 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3993 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/1161 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WTF/wtf/text/Base64.cpp:
```diff
@@ -193,9 +193,9 @@ template<typename T> static std::optional<Vector<uint8_t>> base64DecodeInternal(
                 destination[destinationLength++] = decodedCharacter;
-            } else if (!options.contains(Base64DecodeOptions::IgnoreSpacesAndNewLines) || (!isLatin1(ch) || !isASCIISpace(ch) || (options.contains(Base64DecodeOptions::DiscardVerticalTab) && ch == '\v'))) {
+            // isASCIISpace includes U+000B which https://infra.spec.whatwg.org/#forgiving-base64 excludes
+            } else if (mode != Base64DecodeMode::DefaultIgnoreWhitespace || (!isLatin1(ch) || !isASCIISpace(ch) || ch == '\v'))
```
I wasn't sure how to use isHTMLSpace here.

Also, why is the `!isLatin1(ch)` check here? Performance?

annevk replied:
https://bugs.webkit.org/show_bug.cgi?id=184529#c11 has some context. Given the CSP and SRI failure I should probably reduce the scope of this somewhat, though those failures should be investigated as well as they are shared with Firefox and the specifications are lacking.

annevk replied:
I have reduced the scope now by not enforcing padding validation for all non-URL cases. I will follow-up separately on CSP/SRI as there are standards-related issues there.

darinadler commented on Source/WTF/wtf/text/Base64.cpp:
```diff
@@ -193,9 +194,9 @@ template<typename T> static std::optional<Vector<uint8_t>> base64DecodeInternal(
                 destination[destinationLength++] = decodedCharacter;
-            } else if (!options.contains(Base64DecodeOptions::IgnoreSpacesAndNewLines) || (!isLatin1(ch) || !isASCIISpace(ch) || (options.contains(Base64DecodeOptions::DiscardVerticalTab) && ch == '\v'))) {
+            // isASCIISpace includes U+000B which https://infra.spec.whatwg.org/#forgiving-base64 excludes
+            } else if (mode != Base64DecodeMode::DefaultValidatePaddingAndIgnoreWhitespace || !isASCIISpace(ch) || ch == '\v')
```
Seems a tiny bit sad to have to write this out like this. I wish we could take the infrastructure specification’s version of whitespace and put a sensibly named function somewhere in `WTF` for it. Then `isHTMLSpace` could just call it, or even eventually be removed.

darinadler replied:
I’ll go a step further and say that we should factor this expression into an inline function for this definition of whitespace, even if the function is local to this file. Call it `isWebPlatformWhitespace` or `isInfraWhitespace`? Follows my principle of “use the fact that you had to write a comment to inform how you write the code” and in some cases one can even remove a comment entirely if you name carefully.

darinadler replied:
One last thought. I find that making the correct local function makes it easier to refactor and share later.

annevk replied:
Thanks @darinadler! If #13077 looks good we can land that first and then I will rebase this.

annevk committed:
Combine Base64DecodeOptions and Base64DecodeMode
https://bugs.webkit.org/show_bug.cgi?id=255479
rdar://108082420

Reviewed by Darin Adler.

https://bugs.webkit.org/show_bug.cgi?id=184529#c11 gave a bunch of feedback that was unfortunately
overlooked. This attempts to address it. In particular:

* Using WebKit's isUnicodeCompatibleASCIIWhitespace was an accident and not intentional. All callers
  want isASCIIWhitespace (which excludes U+000B (\v)).
* The isLatin1() check was thought to be needed to address a test, but that wasn't actually the
  case.

Additionally this folds Base64DecodeOptions into Base64DecodeMode reducing the number of decoding
variants we offer to callers and making them more explicit.

We might be able to shrink that even more with some additional investigation, in particular as part
of properly standardizing CSP and SRI base64/base64url requirements.

Base64EncodeMap is renamed to Base64EncodeMode for parity.

* Source/JavaScriptCore/jsc.cpp:
(JSC_DEFINE_HOST_FUNCTION):
* Source/WTF/wtf/text/Base64.cpp:
(WTF::base64EncodeInternal):
(WTF::base64Encode):
(WTF::base64EncodeToVector):
(WTF::base64EncodeToString):
(WTF::base64DecodeInternal):
(WTF::base64Decode):
* Source/WTF/wtf/text/Base64.h:
(WTF::base64EncodeToVector):
(WTF::base64EncodeToString):
(WTF::base64Decode):
(WTF::base64URLEncodeToVector):
(WTF::base64URLEncodeToString):
(WTF::base64URLDecode):
(WTF::calculateBase64EncodedSize):
(WTF::base64Encoded):
(WTF::base64URLEncoded):
(WTF::StringTypeAdapter<Base64Specification>::writeTo const):
* Source/WebCore/page/Base64Utilities.cpp:
(WebCore::Base64Utilities::atob):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::userStyleSheetLocationChanged):

https://bugs.webkit.org/show_bug.cgi?id=255442 covers cleaning this up further, but this at least
aligns it with data: URL requirements.

* Source/WebCore/platform/graphics/avfoundation/CDMFairPlayStreaming.cpp:
(WebCore::extractSinfData):
* Source/WebCore/platform/network/DataURLDecoder.cpp:
(WebCore::DataURLDecoder::decodeSynchronously):
* Source/WebKit/UIProcess/Inspector/gtk/RemoteWebInspectorUIProxyGtk.cpp:
(WebKit::RemoteWebInspectorUIProxy::platformSave):
* Source/WebKit/UIProcess/Inspector/gtk/WebInspectorUIProxyGtk.cpp:
(WebKit::WebInspectorUIProxy::platformSave):
* Source/WebKit/UIProcess/Inspector/mac/WebInspectorUIProxyMac.mm:
(WebKit::WebInspectorUIProxy::showSavePanel):
* Source/WebKitLegacy/mac/WebCoreSupport/WebInspectorClient.mm:
(WebInspectorFrontendClient::save):

Canonical link: https://commits.webkit.org/263298@main
